File: src/main/java/org/example/eshop/repository/OrderItemRepository.java
package org.example.eshop.repository;

import org.example.eshop.entity.OrderItem;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {

    List<OrderItem> findByOrderId(Long orderId);

    List<OrderItem> findByVariantId(Long variantId);

    /**
     * May return null when no item exists for the given orderId and variantId.
     */
    @Nullable
    OrderItem findByOrderIdAndVariantId(Long orderId, Long variantId);

    @Query("SELECT oi FROM OrderItem oi WHERE oi.orderId = :orderId")
    List<OrderItem> findAllByOrderId(@Param("orderId") Long orderId);

    @Query("SELECT SUM(oi.qty) FROM OrderItem oi WHERE oi.variantId = :variantId")
    Long getTotalQuantityOrderedForVariant(@Param("variantId") Long variantId);

    @Query("SELECT oi FROM OrderItem oi JOIN oi.order o WHERE o.paymentStatus = 'PAID' AND oi.variantId = :variantId")
    List<OrderItem> findPaidOrderItemsByVariantId(@Param("variantId") Long variantId);
}

File: src/main/kotlin/org/example/eshop/repository/OrderItemRepository.kt
package org.example.eshop.repository

import org.example.eshop.entity.OrderItem
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface OrderItemRepository : JpaRepository<OrderItem, Long> {
    
    fun findByOrderId(orderId: Long): List<OrderItem>
    
    fun findByVariantId(variantId: Long): List<OrderItem>
    
    fun findByOrderIdAndVariantId(orderId: Long, variantId: Long): OrderItem?
    
    @Query("SELECT oi FROM OrderItem oi WHERE oi.orderId = :orderId")
    fun findAllByOrderId(@Param("orderId") orderId: Long): List<OrderItem>
    
    @Query("SELECT SUM(oi.qty) FROM OrderItem oi WHERE oi.variantId = :variantId")
    fun getTotalQuantityOrderedForVariant(@Param("variantId") variantId: Long): Long?
    
    @Query("SELECT oi FROM OrderItem oi JOIN oi.order o WHERE o.paymentStatus = 'PAID' AND oi.variantId = :variantId")
    fun findPaidOrderItemsByVariantId(@Param("variantId") variantId: Long): List<OrderItem>
}